pkg/parser/gateway_sni_checker.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"fmt"
+	"strings"
+
+	networkingv1alpha3 "istio.io/api/networking/v1alpha3"
+	istioconfig "istio.io/istio/pkg/config"
+	"istio.io/pkg/log"
+)
+
+// this function checks: whether gateway has relaxed sni configuration and if so, do we have the virtual service
+// to prevent insecure access.
+func scanGatewayRelaxedSniHost(inputConfig []configCollection) []error {
+	var gatewayConfigs configCollection
+	var vsConfigs configCollection
+	if len(inputConfig) > 0 {
+		gatewayConfigs = inputConfig[0]
+	}
+	out := []error{}
+	gateways := []gatewayRelaxedSniCheckInfo{}
+	for _, gateway := range gatewayConfigs {
+		gw := findGatewayHostsForRelaxedSniHost(gateway)
+		if gw.relaxedHostGateway != "" {
+			gateways = append(gateways, gw)
+		}
+	}
+	if len(gateways) == 0 {
+		return nil
+	}
+	var filteredGWs = []gatewayRelaxedSniCheckInfo{}
+	for _, gateway := range gatewayConfigs {
+		filteredGWs = append(filteredGWs, checkGWHostsWithHigherTLS(gateway, gateways)...)
+	}
+	if len(filteredGWs) == 0 {
+		return nil
+	}
+
+	if len(inputConfig) > 1 {
+		vsConfigs = inputConfig[1]
+	}
+
+	if len(vsConfigs) > 0 {
+		err := hasVSRejectRelaxedTLSMode(vsConfigs, filteredGWs)
+		if len(err) > 0 {
+			out = append(out, err...)
+		}
+	} else {
+		for _, gtw := range filteredGWs {
+			for _, host := range gtw.hosts {
+				msg := fmt.Errorf("no virtual service configured for gateway %s, at host %s, which is creating problem in gateway:%s, to reject call for host %s", gtw.relaxedHostGateway, host.relaxedHost, gtw.privilegedHostGateway, host.privilegedHost)
```

Comment:
I'd rather use `q` as you never know if there is a `space` typo or something like that when you see the error.

```suggestion
				msg := fmt.Errorf("no virtual service configured for gateway %q, at host %q, which is creating problem in gateway: %q, to reject call for host %q", gtw.relaxedHostGateway, host.relaxedHost, gtw.privilegedHostGateway, host.privilegedHost)
```

Reply:
Done

Reply:
Did you push your changes?